### ISSUE TITLE
Performance of Llama 3.0 8B matches Huggingface fork

### ISSUE THREAD

**Issue by tengyifei** [AI-Hypercomputer/torchprime#16]
We'll need to benchmark `Llama 3.0 8B` at least once to make sure the performance matches [1]. This way we can declare that [1] is deprecated.

[1]: https://github.com/pytorch-tpu/transformers/tree/flash_attention

**Comment by tengyifei** [comment]
We ended up doing some Llama 3.0 runs due to scheduling chaos. Here are some benchmark results (torch_xla pinned to 2.6.0 docker):

Llama 3.0 8B gbs 256 on 1 v6e pod: 1.625s
Llama 3.0 8B gbs 512 on 1 v6e pod: 2.991s

cc @zpcore

**Comment by zpcore** [comment]
Summary of data comparison:

|model config| gbs | tpu| torchprime|huggingface|
|---|---|---|---|--|
|Llama 3.0 8B 8kseq_len| 256| v6e-256 x 1pod | 1.625s|1.71s|
|Llama 3.0 8B 8kseq_len| 512| v6e-256 x 1pod | 2.991s|3.37s|
|Llama 3.0 70B 4kseq_len | 512| v6e-256 x 1pod| 9.480s | 9.435s|

**Comment by zpcore** [comment]
Tested the loss between HF and TP w/ and w/o minibatch:

<img width="895" alt="Image" src="https://github.com/user-attachments/assets/d1ed7791-dbeb-4816-9afc-4933f1621e0d" />

The loss curve is generally close between all those runs.

**Comment by tengyifei** [comment]
@zpcore what's the global batch size?

**Comment by tengyifei** [comment]
Could you run a longer experiment? And plot the results in log scale? That could reveal subtle bugs.

Maybe compare them over 500 steps?

**Comment by zpcore** [comment]
> what's the global batch size?

Global batch size is 256 for the graph. 

> Could you run a longer experiment? And plot the results in log scale? That could reveal subtle bugs. Maybe compare them over 500 steps?

Given the dataset we have, one epoch may not be able to hold 500 steps. Let me do a try.

**Comment by tengyifei** [comment]
> Given the dataset we have, one epoch may not be able to hold 500 steps. Let me do a try.

Yeah, you might also find that `torchprime` quits after one epoch. However, it should be easy to extend it to train multiple epochs.

**Comment by tengyifei** [comment]
Ok, since we tested the performance at one point, I'll make this as closed for Llama 3.0 8B. I've filed a separate issue to test Llama 3.1 8B perf: https://github.com/AI-Hypercomputer/torchprime/issues/133
